js/data/bucket/symbol_bucket.js

```diff
@@ -184,6 +192,12 @@ class SymbolBucket {
             // It's better to place labels on one long line than on many short segments.
             this.features = mergeLines(this.features);
         }
+
+        // set up this mapping to allow for constant-time lookups in place()
+        this.featuresByIndex = {};
```

Comment:
This is done in https://github.com/mapbox/mapbox-gl-js/pull/4217.   Tomorrow AM I'll rebase on that, which will allow us to just place a direct reference to `featureProperties` onto each `symbolInstance` as you suggested.